syntax = "proto3";

// API Version 1.0 (RN-007: Versionamento de API)
package chat4all.v1;

import "common.proto";

option java_multiple_files = true;
option java_package = "chat4all.grpc.generated.v1";
option java_outer_classname = "MessagesProto";

// Serviço de Mensagens
service MessageService {
  // Enviar mensagem para uma conversa
  rpc SendMessage(SendMessageRequest) returns (SendMessageResponse);
  
  // Obter histórico de mensagens
  rpc GetMessages(GetMessagesRequest) returns (GetMessagesResponse);
  
  // Marcar mensagem como lida
  rpc MarkAsRead(MarkAsReadRequest) returns (MarkAsReadResponse);
  
  // Streaming bidirecional para chat em tempo real
  rpc StreamMessages(stream SendMessageRequest) returns (stream MessageNotification);
}

message SendMessageRequest {
  string conversation_id = 1;
  string content = 2;
  string file_id = 3; // Opcional - ID do arquivo previamente uploadado
}

message SendMessageResponse {
  string message_id = 1;
  string conversation_id = 2;
  string status = 3; // ACCEPTED, DELIVERED, READ
  int64 timestamp = 4;
}

message GetMessagesRequest {
  string conversation_id = 1;
  int32 limit = 2;
  int64 offset = 3;
}

message GetMessagesResponse {
  repeated Message messages = 1;
  Pagination pagination = 2;
  string conversation_id = 3;
}

message MarkAsReadRequest {
  string message_id = 1;
  string user_id = 2;
}

message MarkAsReadResponse {
  bool success = 1;
  string message = 2;
}

message Message {
  string message_id = 1;
  string conversation_id = 2;
  string sender_id = 3;
  string content = 4;
  string status = 5;
  int64 timestamp = 6;
  string file_id = 7;
  int64 created_at = 8;
}

message MessageNotification {
  string type = 1; // NEW_MESSAGE, MESSAGE_READ, TYPING, etc
  Message message = 2;
  int64 timestamp = 3;
}
